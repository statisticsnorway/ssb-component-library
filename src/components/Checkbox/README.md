Checkbox
========

> A clickable box selector

### Usage

#### HTML

```html
<div class="ssb-checkbox">
    <input class="page-number" id="option1" value="option1" />
    <label class="checkbox-label" for="option1">Insert label</label>
</div>
```

#### React

```jsx harmony
<Checkbox
  callback={() => callbackFunction()}
  disabled={false}
  index="1"
  selected={false}
  value="option1"
>Item 1
</Checkbox>
```

Available props:

| Name       | Type           | Description  |
| ---------- | ------------- | ----- |
| callback   | func | Function to trigger when value changes |
| children | node | Label for the checkbox |
| disabled | bool | Toggles disabled state of input field |
| index | number | Tab index for focus |
| selected | bool | Selected state of checkbox |
| value | Required number or string | Input field value |
