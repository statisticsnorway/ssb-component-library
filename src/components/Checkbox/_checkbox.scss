.ssb-checkbox {
  @include roboto;
  color: $ssb-dark-6;
  cursor: pointer;
  display: block;
  font-size: 16px;
  min-width: 200px;
  outline: none;
  position: relative;
  text-align: left;
  transition: background .18s;
  user-select: none;

  &:hover, &:focus {
    background: $ssb-green-2;

    .checkbox-label {

      &:before {
        border: 2px solid $ssb-green-4;
      }
    }
  }

  input[type=checkbox] {
    cursor: pointer;
    opacity: 0;
    position: absolute;

    &:checked + .checkbox-label {

      &:before {
        background: $ssb-dark-5;
      }

      &:after {
        border: solid $ssb-white;
        border-width: 0 3px 3px 0;
        content: '';
        display: block;
        height: 12px;
        left: 17px;
        position: absolute;
        top: 14px;
        transform: rotate(45deg);
        width: 6px;
      }
    }

    &:disabled + .checkbox-label {
      color: $ssb-dark-2;

      &:before {
        border: 1px solid $ssb-dark-2;
      }
    }
  }

  .checkbox-label {
    align-items: center;
    cursor: pointer;
    display: flex;
    margin-bottom: 0;
    padding: $global-padding-size;
    position: relative;

    &:before {
      background: $ssb-white;
      border: 1px solid $ssb-dark-5;
      border-radius: 2px;
      content: '';
      display: block;
      height: 20px;
      margin-right: $global-padding-size;
      width: 20px;
    }
  }
}
