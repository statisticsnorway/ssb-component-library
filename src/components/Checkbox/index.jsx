import React from 'react';
import PropTypes from 'prop-types';

const Checkbox = ({
	index, children, callback, disabled, selected, value,
}) => (
	<div className="ssb-checkbox" tabIndex={index + 1}>
		<input
			disabled={disabled}
			id={value}
			checked={selected}
			onChange={() => callback(value)}
			type="checkbox"
			value={value}
		/>
		<label className="checkbox-label" htmlFor={value}>{children}</label>
	</div>
);

Checkbox.defaultProps = {
	callback: () => {},
	index: 1,
};

Checkbox.propTypes = {
	callback: PropTypes.func,
	children: PropTypes.node,
	disabled: PropTypes.bool,
	index: PropTypes.number,
	selected: PropTypes.bool,
	value: PropTypes.oneOfType([
		PropTypes.string,
		PropTypes.number,
	]).isRequired,
};

export default Checkbox;
