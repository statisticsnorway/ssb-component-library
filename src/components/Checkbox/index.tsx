import React, { type ReactNode } from 'react'

interface CheckboxProps {
  callback?: (value: string | number) => void
  children?: ReactNode
  className?: string
  disabled?: boolean
  selected?: boolean
  tabIndex?: number
  value: string | number
}

const Checkbox: React.FC<CheckboxProps> = ({
  callback = () => {},
  children,
  className,
  disabled,
  selected,
  tabIndex = 0,
  value,
}) => (
  <div className={`ssb-checkbox${className ? ` ${className}` : ''}`}>
    <input
      tabIndex={tabIndex}
      disabled={disabled}
      id={value.toString()}
      checked={selected}
      onChange={() => callback(value)}
      type='checkbox'
      value={value}
    />
    <label className='checkbox-label' htmlFor={value.toString()}>
      {children}
    </label>
  </div>
)

export default Checkbox
