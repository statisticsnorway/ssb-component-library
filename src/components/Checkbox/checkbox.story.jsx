import React from 'react'
import Checkbox from './index'

export default {
  title: 'Checkbox',
}

export const Default = () => (
  <div style={{ display: 'flex' }}>
    <Checkbox value='item'>Item</Checkbox>
  </div>
)
