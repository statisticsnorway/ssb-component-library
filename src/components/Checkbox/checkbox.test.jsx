import React from 'react';
import {shallow} from 'enzyme';
import Checkbox from './index';

describe('Checkbox component', () => {
	test('Matches the snapshot', () => {
		const wrapper = shallow(<Checkbox value="item">Item</Checkbox>);
		expect(wrapper).toMatchSnapshot();
	});

	test('Sets selected state by default prop', () => {
		const wrapper = shallow(<Checkbox value="item" selected>Item</Checkbox>);
		expect(wrapper.find('input').props().checked).toEqual(true);
	});

	test('Sends callback', () => {
		const onClick = jest.fn();
		const wrapper = shallow(<Checkbox callback={onClick} value="item">Item</Checkbox>);
		wrapper.find('input').simulate('change', {target: {checked: true}});
		expect(onClick).toBeCalled();
	});
});
