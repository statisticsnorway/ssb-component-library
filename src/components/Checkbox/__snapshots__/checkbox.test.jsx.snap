// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Checkbox component Matches the snapshot 1`] = `
<DocumentFragment>
  <div
    class="ssb-checkbox"
  >
    <input
      id="item"
      tabindex="0"
      type="checkbox"
      value="item"
    />
    <label
      class="checkbox-label"
      for="item"
    >
      Item
    </label>
  </div>
</DocumentFragment>
`;
